Protocol Oriented & Generics Tutorial

bartchr808 opened the pull request:
Corresponds to this [JIRA ticket](https://bugs.swift.org/browse/TF-480). I ended up making a single tutorial since generics can closely relate to how you use protocols. Here's some additional notes I made when writing this:

- Links at the top of the Colab with the TF logo, Colab logo, and GitHub logo are broken right now since this is a PR
- The code is indented 2 spaces but I noticed some Colab walkthroughs we have are 4…should it all be 4?
- It was hard for me to succinctly talk about protocols in Swift for TensorFlow. I thought briefly mentioning a tiny example and referring the user to the documentation and other tutorials would be sufficient. I'm on the fence about removing the last section on Swift for TensorFlow and protocols since it doesn't add much. However, I'd be open to adding more if someone could point me in the right direction on what I could add on!

rxwei commented:
Looks great overall! Thanks a lot for putting this together.

> Links at the top of the Colab with the TF logo, Colab logo, and GitHub logo are broken right now since this is a PR

It's okay. We'll need to set up tutorial images on tensorflow.org/swift internally.

> The code is indented 2 spaces but I noticed some Colab walkthroughs we have are 4…should it all be 4?

Yep, it should all be 4.

> It was hard for me to succinctly talk about protocols in Swift for TensorFlow. I thought briefly mentioning a tiny example and referring the user to the documentation and other tutorials would be sufficient. I'm on the fence about removing the last section on Swift for TensorFlow and protocols since it doesn't add much. However, I'd be open to adding more if someone could point me in the right direction on what I could add on!

I'm pretty swamped recently preparing for I/O, but @dan-zheng will help take a look!

saeta commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
```diff
@@ -0,0 +1,1141 @@
+{
```
<p>Consider replacing the 1st paragraph with:</p><p><br></p><p>OOP has been taught in school as the de-facto design pattern to be used throughout the development stack, from web frontend to mobile to backend.</p><p><br></p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

saeta replied:
<p>This section is set up as a negative. (Basically, we are telling the reader that what they know and love is wrong.) Is there a way we can take the ideas you have in this section, and make them positive? Maybe a narrative arc along the lines of:</p><ul><li>Swift has value types.</li><li>What are value types?</li><li>Why are value types awesome?</li></ul><p>And then include these advantages in that section.</p><p><br></p><p>Does that make sense?</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

burmako replied:
<p>What is protocol-oriented programming? I think it would be useful to introduce this upfront to be able to highlight how the distinguishing features of protocol-oriented programming help writing quantifiably better software (more concise code, better performing code, etc).</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

burmako replied:
<p>Perhaps this tutorial can introduce protocols without starting with shortcomings in alternative approaches? Given that OOP may mean different things for different people and different languages, this may invite controversy which I think can be avoided. An alternative to the current wording would be skipping this section altogether, explaining how protocols work on compelling examples and then discussing pros and cons in comparison with designs that don't involve protocols.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

burmako replied:
<p>Similarly to the previous section, I think that the potential controversy of comparing Swift with everyone's favorite languages can be avoided by saying something like "In addition to classes which have reference semantics, Swift supports enums and structs that are passed by value. Enums and structs support many features provided by classes, albeit with a few restrictions. Let's take a look".</p><p><br></p><p>I believe that Swift's features are deep enough and useful enough to stand on their own, without necessarily referring to alternatives which are supposedly doing worse in some regards. The latter may make the message more powerful, but it also invites further questions and imposes an additional burden of proof, which both make the task of writing more complicated.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

burmako replied:
<p>What's a protocol? I don't think the notebook has introduced this term before its use here.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

burmako replied:
<p>From this sentence, it is unclear why Apartment should inherit from House. For example, what are the pros and cons with respect to Apartment having a House as a property?</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

burmako replied:
<p>Given that a popular use of inheritance is to represent an "is-a" relationship, inheriting from both House and Lawn looks surprising, since it's hard to imagine something that is both a house and a lawn. (Same for a house and an apartment from the code snippets above).</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

burmako replied:
<p>Why not define doIHaveALawn as a method in the House protocol? The tutorial doesn't say that protocols cannot have concrete methods. Also, why can't they have concrete methods?</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

burmako replied:
<p>1) Would adding `doIHaveALawn` to the base class introduce any additional state?</p><p><br></p><p>2) Also, am I right in thinking that overriding via extensions doesn't support dynamic dispatch? I think it would be good to point this out.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

burmako replied:
<p>What does it mean for a type to be built in? Also, is it essential that these types are structs? E.g. are extensions only available for structs?</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

burmako replied:
<p>1) What does "very expressive" mean? How does one quantify expressiveness?</p><p><br></p><p>2) Would "the combined power of protocol oriented programming and generics" change meaning if it said "the combined power of protocols and generics"? If no, what is the difference between protocol-oriented programming and simply using protocols?</p><p><br></p><p>3) Which of these options at our disposal is/are (the most) idiomatic? It would be good to provide some guidance to promote consistency in future S4TF codebases.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

burmako replied:
<p>What are named types and compound types? I don't think they were introduced in this notebook or in the Swift Tour.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='30'/>

burmako replied:
<p>I believe that this wording can be potentially controversial. What if this is the approach that the reader has come up with in the first place? Does this mean that they are now naive? In general, this is another example of why I think that coming up with strawmen and then defeating them to promote a certain narrative may be not the most effective way to communicate.</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

burmako replied:
<p>Why not use a function type instead of predicate structs?</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

burmako replied:
<p>1) It would be good to define what's a protocol-oriented way, so that the reader understands what exactly we find desirable about this way of developing software.</p><p><br></p><p>2) Also, how does one quantify abstraction? Just like with "very expressive", "high level of abstraction" lacks established quantification criteria and therefore may invite controversy, e.g.: "In my favorite language, it is impossible to do this. Does this now mean that my favorite language has a low level of abstraction (i.e. is "bad")?"</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

burmako replied:
<p>Why is it important for the reader to know that protocols are pervasive in S4TF? Will this knowledge (in comparison with the knowledge that you can extend something called Layer whatever this is) help them develop better models? Or the goal is to make the reader feel good about S4TF because it uses advanced engineering techniques such as protocol-oriented programming?</p>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

saeta commented:
Thanks @bartchr808 for this. Super excited to see it go in!

bartchr808 commented:
Thanks for the feedback @rxwei and @saeta ! And I'll be sure to create the google doc tomorrow morning before my orientation @dan-zheng . I also got some additional feedback from a friend which I'll be sure to implement in a commit.

Flew in yesterday to Mountain View so have been swamped with getting settled! I'll be sure to work on this after work this week since orientation is every day I think.

bartchr808 commented:
I sent invites to you @dan-zheng for the google docs to your google email. I also added @rxwei as well. I wanted to invite you @saeta as well if you wanted to take a look but couldn't find your google email.

saeta commented:
Hey @bartchr808 ! My Google email is `saeta@google.com`, but if you could also add `tf-swift-team@google.com` (not sure if it'll work), that will get the whole team too. If not, you can do it this afternoon. :-) Welcome to the team! -Brennan

bartchr808 commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
```diff
@@ -0,0 +1,1108 @@
+{
```
<p><strong>Brennan: </strong></p><blockquote>"Hmm, this section reads both relatively dry and potentially leaves people getting hung up on Swift limitations (e.g. that C++ doesn't have).&nbsp;</blockquote><blockquote><br></blockquote><blockquote>What do you think about this silly idea: what if you illustrated things by example such that the user themselves comes to the conclusion that subclassing / object orientation is inappropriate for certain situations? You'll definitely have to figure out a great example, but I think showing (not telling) might be more powerful here. Does that make sense?"</blockquote><p><br></p><p><strong>Bart:</strong></p><blockquote>"Hmm I'll definitely try figuring something out! My worry is that I may introduce some bias and throw object oriented under the bus, i.e protocol oriented is superior and object oriented sucks, which my previous version of the tutorial sometimes did and Eugene&nbsp;helped correct."</blockquote>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

bartchr808 replied:
<p><strong>Brennan:</strong></p><blockquote>"s/giveHint/hint/ ?"</blockquote><p><br></p><p><strong>Bart:</strong></p><blockquote>"Oh like rename this to something else like `var s` or `var hint`?</blockquote>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

bartchr808 replied:
<p><strong>Brennan:</strong></p><blockquote>var description: String ?</blockquote><blockquote><br></blockquote><blockquote>Maybe change this to "makeNoise" or something to communicate the side-effect nature?</blockquote><p><strong>Bart:</strong></p><blockquote>It doesn't print and returns a string without modifying the struct, so what would the side effect be?<span class="ql-cursor">﻿</span></blockquote>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

bartchr808 commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
```diff
@@ -0,0 +1,1108 @@
+{
```
<p>(regarding `FilterPredicate`)</p><p><strong>Eugene:</strong></p><blockquote>Why not use a function type instead?</blockquote><p><strong>Bart:</strong></p><blockquote>That's definitely another option! I thought of using another protocol, but it can be a bit overkill here.&nbsp;</blockquote><blockquote><br></blockquote><blockquote>Let me know what you think, as to whether just have a function, keep it with the protocol, or show both examples.</blockquote>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

bartchr808 replied:
@burmako Wanted to see what you think

bartchr808 commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
```diff
@@ -0,0 +1,1108 @@
+{
```
<p><strong>Eugene:</strong></p><blockquote>Why is it important for the reader to know that protocols are pervasive in S4TF? Will this knowledge (in comparison with the knowledge that you can extend something called Layer whatever this is) help them develop better models? Or the goal is to make the reader feel good about S4TF because it uses advanced engineering techniques such as protocol-oriented programming?</blockquote><p><strong>Bart:</strong></p><blockquote>Ya I wasn't sure how I feel about this section when I wrote it. I mentioned in the original PR&nbsp;<a href="https://www.google.com/url?q=https://github.com/tensorflow/swift/pull/187&amp;sa=D&amp;ust=1558631981042000&amp;usg=AFQjCNGRTR6U8rx-kpbj-W48PsPodk6-GA" target="_blank">https://github.com/tensorflow/swift/pull/187</a>&nbsp;that I'm on the fence as to whether this section is necessary.</blockquote><blockquote><br></blockquote><blockquote>I thought it would have been nice to add something here to tie why we are talking about protocols in the Swift for TensorFlow repo/project, but similarly when we introduce Swift in the "Intro to Swift" Notebook, we don't tie it back to TensorFlow in any way.</blockquote><blockquote><br></blockquote><blockquote>So let me know what you think about completely removing this section as I don't see it adding much value.</blockquote>

 _Reply via <a href='https://app.reviewnb.com/tensorflow/swift/pull/187/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

bartchr808 replied:
Would be open to any other ideas before I remove this section since it currently isn't that helpful

bartchr808 commented:
I updated the PR which has all of the changes from the Google Doc, but it has a few open questions that still need to be addressed, so I added them here

rxwei commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
````diff
@@ -0,0 +1,1307 @@
+        "Let's say we want to define a `GeometricVector` protocol on floating-point numbers to use in some geometry app we are making which defines 3 important vector operations:\n",
+        "\n",
+        "```swift\n",
+        "protocol GeometricVector where Scalar: FloatingPoint {\n",
````
Move the constraint in the where clause to the declaration of the associated type scalar.

rxwei commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
````diff
@@ -0,0 +1,1307 @@
+        "\n",
+        "```swift\n",
+        "protocol GeometricVector where Scalar: FloatingPoint {\n",
+        "    associatedtype Scalar\n",
````
```suggestion
        "    associatedtype Scalar: FloatingPoint\n",
```

rxwei commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
````diff
@@ -0,0 +1,1307 @@
+        "Let's say we want to store the dimensions of the vector, which the `SIMD` protocol can help us with, so we will make our new type refine the `SIMD` protocol. `SIMD` vectors can be thought of as fixed size vectors that are very fast when you use them to perform vector operations:\n",
+        "\n",
+        "```swift\n",
+        "protocol GeometricVector: SIMD where Scalar: FloatingPoint {\n",
````
Move the `Scalar` constraint to the associated type declaration as well.

rxwei commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
```diff
@@ -0,0 +1,1307 @@
+        "    case green = \"green\"\n",
+        "    case blue = \"blue\"\n",
+        "    // A computed property. Note that enums cannot contain stored properties.\n",
+        "    var giveHint: String {\n",
```
```suggestion
        "    var hint: String {\n",
```

rxwei commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
```diff
@@ -0,0 +1,1307 @@
+        "// Can extend the enum as well!\n",
+        "extension Color {\n",
+        "    // A function.\n",
+        "    func giveHintFunc() -> String {\n",
```
```suggestion
        "    func hint() -> String {\n",
```

bartchr808 replied:
This is an extension on `Color` so will rename it to `hintFunc` instead since we already have the `hint` computed property.

rxwei commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
```diff
@@ -0,0 +1,1307 @@
+        "extension Color {\n",
+        "    // A function.\n",
+        "    func giveHintFunc() -> String {\n",
+        "        return self.giveHint\n",
```
```suggestion
        "        return self.hint\n",
```

rxwei commented on docs/site/tutorials/protocol_oriented_generics.ipynb:
```diff
@@ -0,0 +1,1307 @@
+        "}\n",
+        "\n",
+        "let c = Color.red\n",
+        "print(\"Give me a hint for c: \\(c.giveHintFunc())\")\n",
```
```suggestion
        "print(\"Give me a hint for c: \\(c.hint())\")\n",
```

bartchr808 commented:
Awesome thanks!
